app/src/main/java/com/termux/filepicker/TermuxFileReceiverActivity.java

```diff
@@ -0,0 +1,202 @@
+        final String action = intent.getAction();
+        final String type = intent.getType();
+        final String scheme = intent.getScheme();
+
+        if (Intent.ACTION_SEND.equals(action) && type != null) {
+            final String sharedText = intent.getStringExtra(Intent.EXTRA_TEXT);
+            final Uri sharedUri = intent.getParcelableExtra(Intent.EXTRA_STREAM);
+
+            if (sharedText != null) {
+                if (isSharedTextAnUrl(sharedText)) {
+                    handleUrlAndFinish(sharedText);
+                } else {
+                    String subject = intent.getStringExtra(Intent.EXTRA_SUBJECT);
+                    if (subject == null) subject = intent.getStringExtra(Intent.EXTRA_TITLE);
+                    if (subject != null) subject += ".txt";
+                    promptNameAndSave(new ByteArrayInputStream(sharedText.getBytes(StandardCharsets.UTF_8)), subject);
+                }
+            } else if (sharedUri != null) {
+                handleContentUri(sharedUri, intent.getStringExtra(Intent.EXTRA_TITLE));
+            } else {
+                showErrorDialogAndQuit("Send action without content - nothing to save.");
+            }
+        } else if ("content".equals(scheme)) {
+            handleContentUri(intent.getData(), intent.getStringExtra(Intent.EXTRA_TITLE));
+        } else if ("file".equals(scheme)) {
+            // When e.g. clicking on a downloaded apk:
+            String path = intent.getData().getPath();
+            File file = new File(path);
+            try {
+                FileInputStream in = new FileInputStream(file);
+                promptNameAndSave(in, file.getName());
+            } catch (FileNotFoundException e) {
+                showErrorDialogAndQuit("Cannot open file: " + e.getMessage() + ".");
+            }
+        } else {
+            showErrorDialogAndQuit("Unable to receive any file or URL.");
+        }
+    }
+
+    void showErrorDialogAndQuit(String message) {
+        mFinishOnDismissNameDialog = false;
+        new AlertDialog.Builder(this).setMessage(message).setOnDismissListener(dialog -> finish()).setPositiveButton(android.R.string.ok, (dialog, which) -> finish()).show();
+    }
+
+    void handleContentUri(final Uri uri, String subjectFromIntent) {
+        try {
+            String attachmentFileName = null;
+
+            String[] projection = new String[]{OpenableColumns.DISPLAY_NAME};
+            try (Cursor c = getContentResolver().query(uri, projection, null, null, null)) {
+                if (c != null && c.moveToFirst()) {
+                    final int fileNameColumnId = c.getColumnIndex(OpenableColumns.DISPLAY_NAME);
+                    if (fileNameColumnId >= 0) attachmentFileName = c.getString(fileNameColumnId);
+                }
+            }
+
+            if (attachmentFileName == null) attachmentFileName = subjectFromIntent;
+
+            InputStream in = getContentResolver().openInputStream(uri);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *in*. Execution paths that do not contain closure statements were detected. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).